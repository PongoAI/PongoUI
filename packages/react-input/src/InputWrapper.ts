export * from './components/InputWrapper/index';
