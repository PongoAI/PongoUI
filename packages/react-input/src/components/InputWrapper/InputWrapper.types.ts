import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type InputWrapperSlots = {
  /**
   * The root of the InputWrapper.
   */
  root: Slot<'span'>;

  /**
   * The border around the inner Input element.
   */
  border: NonNullable<Slot<'span'>>;

  /**
   * The inner nearest content to render within the Input.
   */
  contentBefore?: Slot<'span'>;

  /**
   * The inner furthest content to render within the Input.
   */
  contentAfter?: Slot<'span'>;

  /**
   * The helper text slot rendered below the Input.
   */
  helperText?: Slot<'p'>;

  /**
   * The visible label text for the TextField component.
   */
  label?: Slot<'label'>;
};

export type InputWrapperCommons = {
  /**
   * Whether the Input should be disabled.
   *
   * @defaultValue false
   */
  disabled?: boolean;

  /**
   * Whether the Input should be in a danger state.
   *
   * @defaultValue false
   */
  danger?: boolean;

  /**
   * The appearance variant to use for the Input.
   *
   * @defaultValue outline
   */
  appearance?: 'outline' | 'standard' | 'filled';

  /**
   * The size of the Input control.
   *
   * @defaultValue medium
   */
  size?: 'small' | 'medium' | 'large';

  /**
   * The ID used for the helperText element.
   */
  helperTextId?: string;

  /**
   * The ID used for the label element.
   */
  labelId?: string;

  /**
   * The currentValue of the input element. This is used to ensure that the Label transition are correct.
   */
  currentValue?: string;
};

export type InputWrapperProps = ComponentProps<Partial<InputWrapperSlots>> & InputWrapperCommons;

export type InputWrapperState = ComponentState<InputWrapperSlots> & InputWrapperCommons;
