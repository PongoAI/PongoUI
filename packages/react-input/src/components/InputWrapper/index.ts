export * from './InputWrapper';
export * from './InputWrapper.types';
export * from './renderInputWrapper';
export * from './useInputWrapper';
export * from './useInputWrapperStyles';
