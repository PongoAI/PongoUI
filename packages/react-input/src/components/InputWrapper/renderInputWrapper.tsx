import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { InputWrapperSlots, InputWrapperState } from './InputWrapper.types';

export const renderInputWrapper = (state: InputWrapperState) => {
  const { slots, slotProps } = getSlots<InputWrapperSlots>(state);

  return (
    <slots.root {...slotProps.root}>
      {slots.contentBefore && <slots.contentBefore {...slotProps.contentBefore} />}
      {slots.label && <slots.label {...slotProps.label} />}
      {state.root.children}
      {slots.contentAfter && <slots.contentAfter {...slotProps.contentAfter} />}
    </slots.root>
  );
};
