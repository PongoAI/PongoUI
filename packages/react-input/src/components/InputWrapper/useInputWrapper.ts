import * as React from 'react';
import { getNativeElementProps, resolveShorthand } from '@fluentui/react-utilities';
import type { InputWrapperProps, InputWrapperState } from './InputWrapper.types';
import { useFocusableGroup } from '@fluentui/react-tabster';

export const useInputWrapper = (props: InputWrapperProps, ref: React.Ref<HTMLElement>): InputWrapperState => {
  const {
    disabled,
    size = 'medium',
    helperText,
    contentAfter,
    contentBefore,
    border,
    appearance = 'outline',
    danger,
    helperTextId,
    labelId,
    label,
    currentValue,
  } = props;

  const groupFocusAttributes = useFocusableGroup({ tabBehavior: 'limitedTrapFocus' });

  const state: InputWrapperState = {
    size,
    appearance,
    disabled,
    danger,
    labelId,
    currentValue,
    helperTextId,
    components: {
      root: 'span',
      contentBefore: 'span',
      contentAfter: 'span',
      border: 'span',
      helperText: 'p',
      label: 'label',
    },
    root: getNativeElementProps('div', {
      ref,
      ...groupFocusAttributes,
      ...props,
    }),
    border: resolveShorthand(border, { required: true }),
    helperText: resolveShorthand(helperText),
    contentAfter: resolveShorthand(contentAfter),
    contentBefore: resolveShorthand(contentBefore),
    label: resolveShorthand(label),
  };

  if (state.helperText) {
    state.helperText.id = helperTextId;
  }

  if (state.label) {
    state.label.htmlFor = labelId;
  }

  return state;
};
