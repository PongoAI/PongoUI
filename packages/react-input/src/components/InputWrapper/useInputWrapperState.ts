import * as React from 'react';
import { useControllableState, useEventCallback } from '@fluentui/react-utilities';
import type { InputWrapperState } from './InputWrapper.types';

export const useInputWrapperState = (state: InputWrapperState) => {
  return state;
};
