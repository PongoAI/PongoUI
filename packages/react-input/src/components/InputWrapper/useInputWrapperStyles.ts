import { shorthands, makeStyles, mergeClasses } from '@griffel/react';
import type { InputWrapperState } from './InputWrapper.types';
import { tokens } from '@pongo-ui/react-theme';

export const rootClassName = 'pongo-input-wrapper';
export const labelClassName = 'pongo-input-label';
export const legendClassName = 'pongo-input-legend';

const labelFocusedSize = '--focused-label-font-size';

export const useRootStyles = makeStyles({
  root: {
    position: 'relative',
    boxSizing: 'border-box',
    display: 'flex',
    alignItems: 'center',
    flexWrap: 'nowrap',
    color: tokens.textColor,
    backgroundColor: tokens.canvasColor,
    fontFamily: tokens.baseFont,
    ...shorthands.margin('0px'),
    ...shorthands.padding('0px'),
    ...shorthands.borderStyle('none'),
  },

  // Appearance
  outline: {
    ':before': {
      content: '""',
      position: 'absolute',
      top: '0px',
      right: '0px',
      bottom: '0px',
      left: '0px',
      userSelect: 'none',
      pointerEvents: 'none',
      ...shorthands.border('1px', 'solid', tokens.inherit),
      ...shorthands.borderRadius(tokens.rounded),
    },
  },

  standard: {
    ':before': {
      content: '""',
      position: 'absolute',
      top: '0px',
      right: '0px',
      bottom: '0px',
      left: '0px',
      userSelect: 'none',
      pointerEvents: 'none',
      ...shorthands.borderBottom('1px', 'solid', tokens.inherit),
    },
  },

  filled: {
    backgroundColor: tokens.inheritForegroundPressed,
    ...shorthands.borderRadius(tokens.rounded, tokens.rounded, tokens.square, tokens.square),

    ':before': {
      content: '""',
      position: 'absolute',
      top: '0px',
      right: '0px',
      bottom: '0px',
      left: '0px',
      userSelect: 'none',
      pointerEvents: 'none',
      ...shorthands.borderBottom('1px', 'solid', tokens.inherit),
    },
  },

  defaultFocus: {
    ':focus-within': {
      ':before': {
        ...shorthands.borderColor(tokens.brand),
        ...shorthands.borderWidth('2px'),
      },
      [`& .${labelClassName}`]: {
        color: tokens.brand,
      },
      [`& .${legendClassName}`]: {
        width: 'auto',
      },
    },
  },

  dangerFocus: {
    ':focus-within': {
      ':before': {
        ...shorthands.borderColor(tokens.danger),
        ...shorthands.borderWidth('2px'),
      },
      [`& .${labelClassName}`]: {
        color: tokens.danger,
      },
    },
  },

  enabled: {
    ':before': {
      transitionProperty: 'border-width',
      transitionDuration: '.1s',
      transitionDelay: 'cubic-bezier(0.33, 0.0, 0.67, 1)',
      boxSizing: 'border-box',
    },
  },

  disabled: {
    ':before': {
      ...shorthands.borderColor(tokens.inheritDisabled),
    },
  },

  disabledDanger: {
    ':before': {
      ...shorthands.borderColor(tokens.dangerDisabled),
    },
  },

  // Size
  small: {
    height: '35px',
    width: '150px',
    fontSize: tokens.fontSize200,
    [labelFocusedSize]: tokens.fontSize100,
  },

  medium: {
    height: '45px',
    width: '200px',
    fontSize: tokens.fontSize300,
    [labelFocusedSize]: tokens.fontSize200,
  },

  large: {
    height: '50px',
    width: '300px',
    fontSize: tokens.fontSize400,
    [labelFocusedSize]: tokens.fontSize300,
  },

  // Danger
  danger: {
    ':before': {
      ...shorthands.borderColor(tokens.danger),
    },
  },
});

const useHelperTextStyles = makeStyles({
  helperText: {
    ...shorthands.padding('5px', '10px'),
    ...shorthands.margin('0px'),
    fontFamily: tokens.baseFont,
    fontSize: tokens.fontSize200,
    textAlign: 'left',
  },
});

const useContentStyles = makeStyles({
  base: {
    boxSizing: 'border-box',
    '> svg': { display: 'block' },
  },

  before: {
    paddingLeft: '10px',
  },

  after: {
    paddingRight: '10px',
  },

  enabled: {
    color: tokens.inherit,
    fill: tokens.inherit,
  },

  danger: {
    color: tokens.danger,
    fill: tokens.danger,
  },

  disabled: {
    color: tokens.inheritDisabled,
    fill: tokens.inheritDisabled,
  },

  disabledDanger: {
    color: tokens.dangerDisabled,
    fill: tokens.dangerDisabled,
  },
});

export const useLabelStyles = makeStyles({
  label: {
    position: 'absolute',
    ...shorthands.margin('0px'),
    ...shorthands.padding('0px', '2px'),
    left: '8px',
    color: tokens.inherit,
    fontFamily: tokens.baseFont,
    whiteSpace: 'nowrap',
    overflowStyle: 'hidden',
    textOverflow: 'ellipsis',
    touchAction: 'none',
    pointerEvents: 'none',
    userSelect: 'none',
    display: 'inline-flex',
    flexShrink: 0,
    justifyContent: 'left',
    alignItems: 'center',
    transitionProperty: 'transform, font-size, top',
    transitionDuration: '.1s',
    transitionDelay: 'cubic-bezier(0.33, 0.0, 0.67, 1)',

    ':before': {
      content: '""',
      position: 'absolute',
      right: '0px',
      top: '5px',
      left: '0px',
      height: '3px',
      userSelect: 'none',
      pointerEvents: 'none',
      zIndex: -1,
    },
  },

  standardActive: {
    ':focus-within': {
      [`& .${labelClassName}`]: {
        transform: 'translateY(-90%)',
        fontSize: `var(${labelFocusedSize})`,
      },
    },
  },

  outlineActive: {
    ':focus-within': {
      [`& .${labelClassName}`]: {
        transform: 'translateY(-150%)',
        fontSize: `var(${labelFocusedSize})`,
      },
    },
  },

  filledActive: {
    ':focus-within': {
      [`& .${labelClassName}`]: {
        transform: 'translateY(-90%)',
        fontSize: `var(${labelFocusedSize})`,
      },
    },
  },

  standard: {
    transform: 'translateY(-90%)',
    fontSize: `var(${labelFocusedSize})`,
  },

  outline: {
    transform: 'translateY(-150%)',
    fontSize: `var(${labelFocusedSize})`,
  },

  filled: {
    transform: 'translateY(-90%)',
    fontSize: `var(${labelFocusedSize})`,
  },

  outlineBackground: {
    ':before': {
      backgroundColor: tokens.canvasColor,
    },
  },
});

export const useInputWrapperStyles = (state: InputWrapperState) => {
  const rootStyles = useRootStyles();
  const contentStyles = useContentStyles();
  const labelStyles = useLabelStyles();
  const helperTextStyles = useHelperTextStyles();

  state.border.className = mergeClasses(
    rootClassName,
    rootStyles.root,
    rootStyles[state.size!],
    rootStyles[state.appearance!],
    state.danger && rootStyles.danger,
    state.disabled ? rootStyles.disabled : rootStyles.enabled,
    !state.disabled && (state.danger ? rootStyles.dangerFocus : rootStyles.defaultFocus),
    !state.disabled &&
      labelStyles[(state.appearance! + 'Active') as 'standardActive' | 'outlineActive' | 'filledActive'],
    state.disabled && state.danger && rootStyles.disabledDanger,
    state.border.className,
  );

  if (state.helperText) {
    state.helperText.className = mergeClasses(
      helperTextStyles.helperText,
      state.disabled ? contentStyles.disabled : contentStyles.enabled,
      state.danger && contentStyles.danger,
      state.disabled && state.danger && contentStyles.disabledDanger,
      state.helperText.className,
    );
  }

  const contentClasses = [
    contentStyles.base,
    state.disabled ? contentStyles.disabled : contentStyles.enabled,
    state.danger && contentStyles.danger,
    state.disabled && state.danger && contentStyles.disabledDanger,
  ];

  if (state.contentBefore) {
    state.contentBefore.className = mergeClasses(
      ...contentClasses,
      contentStyles.before,
      state.contentBefore.className,
    );
  }
  if (state.contentAfter) {
    state.contentAfter.className = mergeClasses(...contentClasses, contentStyles.after, state.contentAfter.className);
  }

  if (state.label) {
    state.label.className = mergeClasses(
      labelClassName,
      labelStyles.label,
      state.disabled ? contentStyles.disabled : contentStyles.enabled,
      state.danger && contentStyles.danger,
      state.disabled && state.danger && contentStyles.disabledDanger,
      (state.currentValue !== (undefined || '') || state.contentBefore !== undefined) &&
        state.label !== undefined &&
        labelStyles[state.appearance!],
      state.appearance === 'outline' && labelStyles.outlineBackground,
      state.label.className,
    );
  }

  return state;
};
