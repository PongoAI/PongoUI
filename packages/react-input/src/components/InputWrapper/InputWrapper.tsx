import * as React from 'react';
import { useInputWrapper } from './useInputWrapper';
import { renderInputWrapper } from './renderInputWrapper';
import { useInputWrapperStyles } from './useInputWrapperStyles';
import type { InputWrapperProps } from './InputWrapper.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * The InputWrapper components sets up the various appearances for the Input related components.
 */
export const InputWrapper: ForwardRefComponent<InputWrapperProps> = React.forwardRef((props, ref) => {
  const state = useInputWrapper(props, ref);

  useInputWrapperStyles(state);

  return renderInputWrapper(state);
});
InputWrapper.displayName = 'InputWrapper';
