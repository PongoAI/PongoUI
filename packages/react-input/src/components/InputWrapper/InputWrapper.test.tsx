import React from 'react';
import { InputWrapper } from './InputWrapper';
import { InputWrapperProps } from './InputWrapper.types';
import { isConformant } from '../../common/isConformant';
import { render } from '@testing-library/react';

describe('InputWrapper', () => {
  isConformant({
    Component: InputWrapper as React.FunctionComponent<InputWrapperProps>,
    displayName: 'InputWrapper',
    disabledTests: ['make-styles-overrides-win'],
  });

  it('renders a basic InputWrapper', () => {
    const renderedComponent = render(<InputWrapper />);

    expect(renderedComponent.baseElement).toMatchSnapshot();
  });

  it('renders a basic InputWrapper danger', () => {
    const renderedComponent = render(<InputWrapper danger />);

    expect(renderedComponent.baseElement).toMatchSnapshot();
  });

  it('renders a basic InputWrapper disabled', () => {
    const renderedComponent = render(<InputWrapper disabled />);

    expect(renderedComponent.baseElement).toMatchSnapshot();
  });
});
