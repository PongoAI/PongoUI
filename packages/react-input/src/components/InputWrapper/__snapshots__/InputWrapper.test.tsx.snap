// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputWrapper renders a basic InputWrapper 1`] = `
<body>
  <div>
    <span
      class="pongo-input-wrapper ___1pqmah3 f10pi13n f1ewtqcl ftuwxu6 fv68sux fp6vxd fa4t5tb flvyvdh fc467gg f8qjfyk fk6fouc f4d9j23 f1lmfglv fibjyge f1abmfm4 f9yszdx f1nbblvp fifp7yv f1ov4xf1 f1asdtw4 f1ern45e f1n71otn f1h8hb77 f1deefiw f1rbvai9 f13ju9t4 f1lh17z1 f1dlw1sp f1arxnym fqzc6w8 fuzzvh5 fxf9f1y frdto5n ff0gctb f3zr33r f8g0ba1 f5uakvk frcipv9 f199l62q f1pa2qi2 fyqop9u f1ooiv2t f1bwkb58 fiio4mc f1m5ya7j fc8z0qo fmw406p f12c27rx f17thq47 f12kdog7 f5qrn4q f1kndqco f1kc1xys f1d7l9b4 f16n8puj fkcy365 f5le7dr f170vdtw fg4eyww f1cngt4b f17x5ctf fbk6y72 f1uajmgh f11t0hz4 f2v3se5 f18saz9o f80sxui f19q3xa0 fzmdvw2 fbh30h8"
      data-tabster="{\\"groupper\\":{\\"tabbability\\":2}}"
      style="--pongo-input-line-height: 45px; --pongo-input-width: 200px;"
    />
  </div>
</body>
`;

exports[`InputWrapper renders a basic InputWrapper danger 1`] = `
<body>
  <div>
    <span
      class="pongo-input-wrapper ___9sgl5f0 f10pi13n f1ewtqcl ftuwxu6 fv68sux fp6vxd fa4t5tb flvyvdh fc467gg f1jay0ds fk6fouc f4d9j23 f1lmfglv fibjyge f1abmfm4 f9yszdx f1nbblvp fifp7yv f1ov4xf1 f1asdtw4 f1ern45e f1n71otn f1h8hb77 f1deefiw f1rbvai9 f13ju9t4 f1lh17z1 f1dlw1sp f1arxnym fqzc6w8 fuzzvh5 fxf9f1y frdto5n ff0gctb f3zr33r f8g0ba1 f5uakvk frcipv9 f199l62q f1pa2qi2 fyqop9u f1ooiv2t f1bwkb58 fiio4mc f1m5ya7j fc8z0qo f1tsuuen f1jyinwm f1gmfdaf ff4t21v f5qrn4q f1kndqco f1kc1xys f1d7l9b4 f16n8puj fkcy365 f5le7dr f170vdtw f1md1kfu f1385xuj fwuljub fophbcs f1uajmgh f11t0hz4 f2v3se5 f18saz9o f4heoct fzmdvw2 fbh30h8"
      data-tabster="{\\"groupper\\":{\\"tabbability\\":2}}"
      style="--pongo-input-line-height: 45px; --pongo-input-width: 200px;"
    />
  </div>
</body>
`;

exports[`InputWrapper renders a basic InputWrapper disabled 1`] = `
<body>
  <div>
    <span
      class="pongo-input-wrapper ___1oeiaag f10pi13n f1ewtqcl ftuwxu6 fv68sux fp6vxd fa4t5tb flvyvdh fc467gg f8qjfyk fk6fouc f4d9j23 f1lmfglv fibjyge f1abmfm4 f9yszdx f1nbblvp fifp7yv f1ov4xf1 f1asdtw4 f1ern45e f1n71otn f1h8hb77 f1deefiw f1rbvai9 f13ju9t4 f1lh17z1 f1dlw1sp f1arxnym fqzc6w8 fuzzvh5 fxf9f1y frdto5n ff0gctb f3zr33r f8g0ba1 f5uakvk frcipv9 f199l62q f1pa2qi2 fyqop9u f1ooiv2t f1bwkb58 fiio4mc f1m5ya7j fc8z0qo flm3n1z f1pltsua f15x2d8x fhngjcl f5qrn4q f1kndqco f1kc1xys f1d7l9b4"
      data-tabster="{\\"groupper\\":{\\"tabbability\\":2}}"
      style="--pongo-input-line-height: 45px; --pongo-input-width: 200px;"
    />
  </div>
</body>
`;

exports[`InputWrapper renders a label InputWrapper 1`] = `
<body>
  <div>
    <span
      class="pongo-input-wrapper ___1pqmah3 f10pi13n f1ewtqcl ftuwxu6 fv68sux fp6vxd fa4t5tb flvyvdh fc467gg f8qjfyk fk6fouc f4d9j23 f1lmfglv fibjyge f1abmfm4 f9yszdx f1nbblvp fifp7yv f1ov4xf1 f1asdtw4 f1ern45e f1n71otn f1h8hb77 f1deefiw f1rbvai9 f13ju9t4 f1lh17z1 f1dlw1sp f1arxnym fqzc6w8 fuzzvh5 fxf9f1y frdto5n ff0gctb f3zr33r f8g0ba1 f5uakvk frcipv9 f199l62q f1pa2qi2 fyqop9u f1ooiv2t f1bwkb58 fiio4mc f1m5ya7j fc8z0qo fmw406p f12c27rx f17thq47 f12kdog7 f5qrn4q f1kndqco f1kc1xys f1d7l9b4 f16n8puj fkcy365 f5le7dr f170vdtw fg4eyww f1cngt4b f17x5ctf fbk6y72 f1uajmgh f11t0hz4 f2v3se5 f18saz9o f80sxui f19q3xa0 fzmdvw2 fbh30h8"
      data-tabster="{\\"groupper\\":{\\"tabbability\\":2}}"
      style="--pongo-input-line-height: 45px; --pongo-input-width: 200px;"
    >
      <label
        class="pongo-input-label ___1ll8q7b f1euv43f f1lmfglv fibjyge f1abmfm4 f9yszdx f1nbblvp f12huiiw f1ov4xf1 f161knb0 fpnytvn f9j3dip f1ysjbbf fk6fouc fz5stix f1q148li f1cmbuwj f1sdsnyy f1aehjj5 f1xqy1su ftuwxu6 fi64zpg fc7pyfw f122n59 fvz7k9s fkfijr0 fsxfru3 fuzzvh5 fxf9f1y ff0gctb f8g0ba1 fsa65no f5uakvk frcipv9 fsm07gi f1qo5dib f188r07x fuv0bne fb322o2"
      >
        test
      </label>
    </span>
  </div>
</body>
`;
